<a href="https://colab.research.google.com/github/allseenn/ab/blob/main/terms.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Термины #


+ population - генеральная совокупность
+ sample - выборка
+ random variable - Случайная величина (ξ)
+ ретенш - коэффициент удержания (клиента), см. customer retention
+ control - контрольная групп А
+ treatment - тестовая группа B

рандомизация - случайная выборка 

#Статистика (Statisics) #
**Statistics** is a form of mathematical analysis that uses quantified models and representations for a given set of experimental data or real-life studies.
+ Type of Analytics
+ Probability
+ Central Tendency
+ Variability
+ Relationship Between Variables
+ Probability Distribution
+ Hypothesis Testing and Statistical Significance
+ Regression


## Type of Analytics ## 
**Descriptive Analytics** *(описательная аналитика)* - рассказывает нам, что произошло в прошлом, и помогает бизнесу понять, как он работает, предоставляя контекст, помогающий заинтересованным сторонам интерпретировать информацию.

**Diagnostic Analytics** *(диагностическая аналитика)* - глубже чем просто описание данных, помогает понять, почему что-то произошло в прошлом. 

**Predictive Analytics** *предсказательная аналитика*- предсказывает, что, скорее всего, произойдет в будущем, и предоставляет компаниям полезную информацию на основе информации.

**Prescriptive Analytics** (предписывающая аналитика) - предоставляет рекомендации относительно действий, которые будут использовать преимущества прогнозов и направлять возможные действия к решению.

## Probability ##
**Probability** Вероятность — это мера возможности того, что событие произойдет в случайном эксперименте.

**Complement** *(Дополнение)*: P(A)+P(A’) = 1

**Intersection** *(Пересечение*): P(A∩B)=P(A)P(B)

**Union** *(Объединиение)*: P(A∪B)=P(A)+P(B)−P(A∩B) 

<img src="https://github.com/allseenn/ab/blob/main/img/union.webp?raw=1" width="500">

**Conditional Probability** (Условная вероятность): P(A|B) - это мера возможности возникновения одного события с некоторым отношением к одному или нескольким другим событиям. P(A|B)=P(A∩B)/P(B), когда P(B)>0.

**Independent Events** *(Независимые события)*: два события независимы, если возникновение одного не влияет на вероятность возникновения другого. P(A∩B)=P(A)P(B), где P(A) != 0 и P(B) != 0, P(A|B)=P(A), P(B|A)=P(B).

**Mutually Exclusive Events** *(Взаимоисключающие (несовместные) события)*: два события взаимоисключающие, если они не могут произойти одновременно. P(A∩B)=0 и P(A∪B)=P(A)+P(B).

**Bayes’ Theorem** *(Теорема bayesa)* $P(A|B) = \frac{P(A\cap B)}{P(B)} = \frac{P(A)\times P(B\mid A)}{P(B)}$

## Central Tendency *(Описательные статистики)* ##

**Mean** / **Mathematical expectation**  Математическое ожидание (μ, M) - это среднее арифметическое значение случайной величины или среднее значение набора данных.
$ \overline{x} = \frac{1}{n} \sum_{i=1}^{n}x_{i} $

**Median** (Медиана): среднее значение упорядоченного набора данных.

**Skewness** (Перекос): мера симметрии.

<img src="https://github.com/allseenn/ab/blob/main/img/skew.png?raw=1" width="600">

**Kurtosis** (Куртоз): мера того, являются ли данные тяжелохвостыми или легкохвостыми относительно нормального распределения.

<img src="https://github.com/allseenn/ab/blob/main/img/kurtic.png?raw=1" width="400">


## Variability *Вариации* ##
**Percentiles** Процентили — мера, которая указывает на значение, ниже которого падает данный процент наблюдений в группе наблюдений.

**Quantiles** Квантили (Q) — значения, которые делят количество точек данных на четыре более или менее равные части или четверти.

**Interquartile Range (IQR)** Межквартильный диапазон (IQR) — мера статистической дисперсии и изменчивости, основанная на делении набора данных на квартили. IQR = Q3−Q1

<img src="https://github.com/allseenn/ab/blob/main/img/box.webp?raw=1" width="600">

**Variance** Дисперсия (S2): средняя квадратичная разница значений от среднего для измерения того, насколько распределен набор данных относительно среднего. $ \sigma_x^2 = \frac{1}{n-1}\sum_{i=1}^n(x_1-\overline{X})^2 $

**Standard Deviation** Стандартное отклонение (SD)(σ): стандартная разница между каждой точкой данных и средним и квадратным корнем дисперсии. $ \sigma_x =\sqrt{\frac{1}{n-1}\sum_{i=1}^n(x_1-\overline{X})^2} $

<img src="https://github.com/allseenn/ab/blob/main/img/var.webp?raw=1" width="400">

**Standard Error(SE)** Стандартная ошибка (SE): оценка стандартного отклонения распределения выборки. $ SE_{\overline{x}}=\frac{\sigma}{\sqrt{n}} $

<img src="https://github.com/allseenn/ab/blob/main/img/error.webp?raw=1" width="400">

## Relationship Between Variables (Связь между переменными) ## 
**Causality** (Причинно-следственная связь): связь между двумя событиями, когда на одно событие влияет другое.

**Covariance** (Ковариация): количественная мера совместной изменчивости между двумя или более переменными.

**Correlation** (Корреляция): Измерьте связь между двумя переменными и диапазонами от -1 до 1, нормализованная версия ковариации.

<img src="https://github.com/allseenn/ab/blob/main/img/cov.webp?raw=1" width="400">

## Probability Distribution *Распределение вероятностей* ##
**Распределение вероятностей** — это закон, описывающий область значений случайной величины и соответствующие вероятности появления этих значений.

<img src="https://github.com/allseenn/ab/blob/main/img/distrs.png?raw=1" width="600">

### Probability Distribution Functions *Функции распределения вероятностей* ###
**Probability Mass Function(PMF)** *Функция массы вероятности*: функция, которая дает вероятность того, что дискретная случайная величина в точности равна некоторому значению.

**Probability Density Function(PDF)** *Функция плотности вероятности*: функция для непрерывных данных, где значение в любой данной выборке может быть интерпретировано как обеспечивающее относительную вероятность того, что значение случайной величины будет равно этой выборке.

**Cumulative Density Function(CDF)** *Интегральная функция плотности*: функция, которая дает вероятность того, что случайная величина меньше или равна определенному значению.

<img src="https://github.com/allseenn/ab/blob/main/img/com.webp?raw=1" width="400">

### Continuous Probability Distribution *Непрерывное распределение вероятностей* ###
**Uniform Distribution** *Равномерное распределение*: также называемое прямоугольным распределением, представляет собой распределение вероятностей, где все результаты одинаково вероятны.

**Normal/Gaussian Distribution** *Нормальное/гауссовское распределение*: кривая распределения имеет колоколообразную форму и симметрична и связана с центральной предельной теоремой о том, что выборочное распределение выборки означает, что приближается к нормальному распределению по мере увеличения размера выборки. $ f(x) = \frac{1}{\sigma\sqrt{2\pi}}exp(-\frac{(x-\mu)^2}{2\sigma^2}) $

<img src="https://github.com/allseenn/ab/blob/main/img/norm.png?raw=1" width="400">

Нормальное распределение самый частый вид, который используется в статистике.


<img src="https://github.com/allseenn/ab/blob/main/img/distr.webp?raw=1" width="800">

**Exponential Distribution** *Экспоненциальное распределение*: распределение вероятности времени между событиями в точечном процессе Пуассона.

**Chi-Square Distribution** *Распределение Хи-Квадрат*: Распределение суммы квадратов стандартных нормальных отклонений.

<img src="https://github.com/allseenn/ab/blob/main/img/exp.webp?raw=1" width="500">

### Discrete Probability Distribution *Дискретное распределение вероятностей* ###
**Bernoulli Distribution** *Распределение Бернулли*: распределение случайной величины, которое принимает одно испытание и только 2 возможных исхода, а именно 1 (успех) с вероятностью p и 0 (неудача) с вероятностью (1-p).

**Binomial Distribution** *Биномиальное распределение*: распределение числа успехов в последовательности n независимых экспериментов, каждый из которых имеет только 2 возможных исхода, а именно 1 (успех) с вероятностью p и 0 (неудача) с вероятностью (1-p).

**Poisson Distribution** *Распределение Пуассона*: распределение, которое выражает вероятность возникновения заданного числа событий k в фиксированном интервале времени, если эти события происходят с известной постоянной средней скоростью λ и независимо от времени.

<img src="https://github.com/allseenn/ab/blob/main/img/bin.webp?raw=1" width="500">




## The central limit theorem (Центральная предельная теорема) ##
**CLT** *ЦПТ* говорит, что если мы возьмем достаточно большую выборку из *независимых одинаково распределенных* случайных величин **н.о.р.** (independent and identically-distributed **i.i.d**) из генеральной совокупности , то среднее значение будет нормально распределено с μ(мат ожидание) и SD(стандартное отклонение).
ЦПТ дает возможность строить доверительные интервалы и проверять статистические гипотезы.

## Confidence intervals (Доверительные интервалы) ##
Это способ оценки метрики, используя который, мы получим диапазон значений [x,y], внутри которого будет лежать истинное значение метрики ГС в 95% случаев.
Eсли провести очень большое количество независимых экспериментов с аналогичным построением доверительного интервала, то в 95% экспериментов доверительный интервал будет содержать оцениваемый параметр ген совокупности.
В оставшихся 5% экспериментов доверительный интервал не будет содержать параметр ген совокупности.
### Confidence interval for proportions (Доверительный интервал для долей) ###
**Доверительный интервал для доли** применим к шкалам любого типа (предпочтительно - к категориальным), т.к. расчёт строится на основе частотных распределений. Сам доверительный интервал показывает, в каких границах находится интересующая доля в генеральной совокупности. Если построить интервалы отдельно для нескольких долей, то на основании них можно судить о наличии или отсутствии статистически значимых различий: если интервалы пересекаются, различий нет, если не пересекаются - различия есть. Такая процедура - аналог z-теста, в котором проверяется гипотеза о равенстве долей 
$ \hat{p}\pm1.96*\sqrt{\frac{\hat{p}(1-\hat{p})}{n}} $
### Confidence interval for median (Доверительный интервал для медианы) ###
Доверительный интервал для медианы применим к шкалам порядкового (рангового) типа и выше. Сам доверительный интервал показывает, в каких границах находится медиана признака в генеральной совокупности. Если построить интервалы отдельно для нескольких медиан, то на основании них можно судить о наличии или отсутствии статистически значимых различий: если интервалы пересекаются, различий нет, если не пересекаются - различия есть. Такая процедура - аналог некоторых непараметрических методов, в которых проверяется гипотеза о равенстве медиан.
```
j: nq – z√ nq(1-q)
k: nq + z√ nq(1-q)
    n: размер выборки
    q: квантиль интереса. В качестве медианы мы будем использовать q = 0,5.
    z: z-критическое значение
```
https://www.codecamp.ru/blog/confidence-interval-for-median/

### Confidence interval for the mean (Доверительный интервал для среднего) ###
Доверительный интервал для среднего применим к шкалам интервального типа и выше. Сам доверительный интервал показывает, в каких границах находится математическое ожидание (среднее арифметическое) признака в генеральной совокупности. Если построить интервалы отдельно для нескольких средних, то на основании них можно судить о наличии или отсутствии статистически значимых различий: если интервалы пересекаются, различий нет, если не пересекаются - различия есть. Такая процедура - аналог t-тестов,
в которых проверяется гипотеза о равенстве математических ожиданий.
```
Доверительный интервал = x +/- z*(s/ √n )

    x : выборочное среднее
    z: выбранное значение z
    s: стандартное отклонение выборки
    n: размер выборки
```
https://www.codecamp.ru/blog/confidence-interval-mean/

### Confidence interval for standard deviation (Доверительный интервал для стандартного отклонения) ###
```
Доверительный интервал = [√(n-1)s 2 /X 2 α/2 , √(n-1)s 2 /X 2 1-α/2 ]

    n: размер выборки
    s: стандартное отклонение выборки
    X 2 : Критическое значение хи-квадрат с n-1 степенями свободы.
```
https://www.codecamp.ru/blog/confidence-interval-standard-deviation/

## Metrics (Метрики) ##

<a href="https://habr.com/ru/post/675722/#Finansovye-metriki">Полный список метрик</a>

## Типы метрик ##
1. Доли - (ретеншн, конверсии) [0,1,0,0,0,1]
2. Непрерывные - таймспент в сек / деньги
3. Отношения - (клики на сессию , поездки на юзера))

### Customer retention (удержание клиента) ###
#### 1. Customer Retention Rate (CRR) Коэффициент удержания клиентов ####
Retention rate показывает, сколько клиентов остались с вами за определенный период времени. Например, можно измерить коэффициент удержания за год, месяц или неделю.
```
    Retention Rate = ((CE — CN) / CS)) x 100
    CE = Количество клиентов на конец периода
    CN = Количество новых клиентов, приобретенных за период
    CS = Количество клиентов на начало периода
```
#### 2. Customer Lifetime Value (LTV) Пожизненная ценность клиента ####
Показатель customer lifetime value — это проекция потенциального дохода, который вы можете получить от клиента. Эта метрика рассчитывается на основании прежнего покупательского поведения клиента — так что не стоит сразу переводить ее в быстрые деньги.
```
Customer Lifetime Value = (Средний чек) x (Количество повторных продаж) x (Среднее время удержания)
```
#### 3. Repeat Purchase Rate (RPR) Коэффициент повторных покупок ####
Repeat purchasing rate — это процент клиентов, которые купили ваши продукты более одного раза. Коэффициент повторных покупок показывает, сколько клиентов возвращаются после первой покупки.
```
Repeat purchase rate = Количество повторных клиентов / Общее число клиентов
```
#### 4. Redemption Rate (RR) Коэффициент использования вознаграждения ####
По коэффициенту redemption можно судить, насколько купон мотивирует пользователей купить продукт. Если redemption rate низкий (около 20%, то нужно копать глубже и разбираться, почему продукт не покупают. Возможно, условия акции недостаточно выгодные — или людям просто больше не интересно ваше предложение.
```
Redemption rate = Количество использованных купонов / Общее число купонов
```
#### 5. Net Promoter Score (NPS) Индекс потребительской лояльности ####
Какова вероятность, что вы порекомендуете этот продукт другу или коллеге?

Это и есть net promoter score. Вы задаете клиентам этот простой вопрос, и они отвечают по шкале от 0 до 10, где 0 означает “точно не порекомендую”, а 10 — ”обязательно порекомендую”. Это очень мощная метрика, потому что клиенты рекомендуют друзьям только тот продукт, которым на 100% довольны.
```
NPS = % промоутеров — % критиков
```
В рамках NPS выделяют три категории клиентов:

Критики (0–6) — это люди, недовольные продуктом или сервисом. Они больше ничего у вас не купят и могут испортить репутацию бренда своими жалобами в СМИ и соц.сетях.

Нейтралы (7–8) — это удовлетворенные клиенты, но они не будут особенно восторгаться вашим брендом и могут даже уйти к конкурентам, если те предложат более выгодные условия.

Промоутеры (9–10) — это те, кто сделают все возможное, чтобы рассказать другим о вашем продукте. Они верны бренду и готовы запустить мощное сарафанное радио, которое поможет вам практически без издержек привлечь новых клиентов.

### Advertising metrics (Рекламные метрики) ###
Impressions – сколько всего раз реклама была показана (повторно одному и тому же человеку в том же числе).

Reach – сколько всего людей посмотрело рекламу (встречается в facebook).

1. **CPC** (Cost Per Click) – сколько стоит один клик по рекламе.

$\frac{\text{сколько потрачено денег на рекламу}}{\text{сколько людей кликнули на рекламу}}$

2. **CTR** (Click-through rate) – конверсия рекламы, то есть отношение числа кликов к числу показов.

$\frac{\text{сколько людей кликнули на рекламу}}{\text{сколько людей увидело рекламу}}\times100$

3. **CPM** (Cost per mille) – cколько стоит 1000 показов рекламного объявления.

$\frac{\text{сколько потрачено денег на рекламу}}{\text{сколько было показов рекламы}}\times1000$

4. **CR** (Conversion rate) – конверсия страницы приложения в магазине (какая часть пользователей, которые перешли на страницу приложения по рекламе, потом скачали приложение). В нашем случае мы считаем, что все пользователи перешли на страницу приложения в магазине благодаря рекламе

$\frac{\text{сколько пользователей скачали приложение}}{\text{сколько людей кликнули на рекламу}}\times100$

5. **CPI** (Cost per Install) – стоимость установки приложения из всего рекламного канала.
Иногда используется термин CPA (Cost per Acquisition)

$\frac{\text{сколько потрачено денег на рекламу}}{\text{сколько пользователей скачали приложение}}$


## Hypothesis Testing and Statistical Significance *(Проверка гипотез и статистическая значимость)* ##
**Статистическая гипотеза** - это выдвигаемое предположение о свойствах и характеристиках генеральной совокупности(ей).
После выдвижения гипотезы собираются реальные данные выборок(выборки) и к ним применяются статистические методы(критерии) например t-test или z-test
После применения критериев мы либо отвергаем, либо не отвергаем гипотезу
### Null and Alternative Hypothesis *(Нулевая и альтернативная гипотеза)* ###
**Null Hypothesis** *Нулевая гипотеза*: принимаемое предположение о том, что не существует связи между наблюдениями в двух (или более) событиях (выборках, феноменах, совокупностях). Гипотезу отвергают, если данные показывают разницу между выборками.
**Hypothesis** *Альтернативная гипотеза*: Противоречить нулевой гипотезе.

Гипотезы могут быть односторонними (right or left-tail test) и двухсторонними (Two-tail test). На практике почти всегда используют двухсторонние тк мы почти никогда не знаем заранее перед запуском теста в какую сторону отклонится метрика в тестовой группе.

<img src="https://github.com/allseenn/ab/blob/main/img/hypo.jpg?raw=1" width="600">

### Interpretation *(Интерпретация)* ###
**P-value** *P-значение*: вероятность того, что тестовая статистика будет по крайней мере такой же экстремальной, как и наблюдаемая, учитывая, что нулевая гипотеза верна. Когда p-value > α, мы не  отвергаем нулевую гипотезу, в то время как p-value ≤ α, мы отвергаем нулевую гипотезу и можем заключить, что у нас есть значимый результат.

**Critical Value** *критическое значение*: точка или точки на шкале тестовой статистики, за которой мы отвергаем нулевую гипотезу, и выводится из уровня значимости α теста. Это зависит от статистики теста, которая специфична для типа теста, и уровня значимости, α, который определяет чувствительность теста.

**Significance Level and Rejection Region** *Статистическая значимость (α) и критическая область*: Площадь критической области зависит от величины статистической значимости. Зтатистическая значимость обозначается α и является вероятностью отклонения нулевой гипотезы, если она верна.

**Area of acceptance** *Областью принятия гипотезы* - называют совокупность значений статистики критерия, при которых гипотезу H0(нулевую) принимают.

<img src="https://github.com/allseenn/ab/blob/main/img/hypo.webp?raw=1" width="600">

### Type I and Type II errors (Ошибки первого и второго рода) ###
**True Positive** = говорим истина, когда по факту истина

**True Negative** = говорим не истина, когда по факту тоже не истина

**False Positive** (ошибка I рода) = говорим истина, когда по факту не истина. Отклонение верной нулевой гипотезы. Риск совершить такую ошибку равен выбранному уровню статистической значимости (например, α=0.05) (ошиблись)

**False Negative** (ошибка II рода) = говорим не истина, когда по факту истина (ошиблись). Принятие неверной нулевой гипотезы. Вероятность ошибочно сохранить неверную нулевую гипотезу обозначают буквой β (например, 0.2).

В статистической проверке гипотез ошибка I рода — это отклонение нулевой гипотезы если она верна, в то время как ошибка II рода — это не отклонение нулевой гипотезы, если она не верна.

<img src="https://github.com/allseenn/ab/blob/main/img/conclude.webp?raw=1" width="600">

## Statistical types | tests (Статистические критерии | тесты) ##
**Статистический критерий(тест)** — это математическое правило, которое помогает нам понять, отвергнута или нет наша гипотеза.
**Power of a test** *Мощность критерия/теста* - это вероятность, что тест правильно засечёт эффект там, где он и правда есть. (т.е. 1-β)

Для того, чтобы проверить гипотезу о равенстве показателей, применяется два типа критериев оценки: параметрические и непараметрические.

**Parametric test** *Параметрическими* называются критерии, в которых мы можем сделать предположение о распределении , относящееся к какой-то выборке. В большинстве случаев в качестве распределения используется нормальное.(t тест)

**Non-parametric test** *Непараметрические* не используют предположения о распределении, а оперируют рангами и частотами.(U-критерий Манна — Уитни )

Часто используются t-test Уэлча для сравнения средних

z - критерий , для сравнения метрик долей(конверсии/ретеншн)

критерий Хи-квадрат тоже подойдет для долей + Критерий Манна Уитни 
для сравнения ненормально распределенных величин.

Критерии долей — это критерии, которые работают с распределениями Бернулли. Они принимают на вход выборки из нулей и единиц и проверяют гипотезы о вероятности появления единицы в выборке

### Z-test *(Z-тест)* ###
**Z-тест** — это любой статистический тест, для которого распределение тестовой статистики по нулевой гипотезе может быть аппроксимировано нормальным распределением и проверяет среднее распределение, в котором мы уже знаем дисперсию популяции. Поэтому многие статистические тесты могут быть удобно выполнены в качестве приблизительных Z-тестов, если размер выборки велик или известна дисперсия популяции.

<img src="https://github.com/allseenn/ab/blob/main/img/ztest.webp?raw=1" width="400">

### T-test *(T-тест)* ###
**Т-тест** является статистическим тестом, если дисперсия популяции неизвестна и размер выборки не велик (n < 30).

**Paired sample** *зависимые выборки* - мы дважды собираем данные от одной и той же группы, человека, предмета или вещи.

**Independent sample** *независимые выборки* - две выборки должны были происходить из двух совершенно разных генеральных совокупностей.

<img src="https://github.com/allseenn/ab/blob/main/img/ttest.webp?raw=1" width="600">

У каждого критерия есть предпосылки - и можно использовать таблицу ниже для выбора подходящего критерия.

<img src="https://github.com/allseenn/ab/blob/main/img/tests.png?raw=1" width="800">

## Алгоритм проверки гипотез ##
    1. Выбираем метрику и формулируем нулевую и альтернативную гипотезы
    2. Выбираем параметр alpha (например 5% ) равный вероятности допустить ошибку первого рода
    3. Выбираем критерий, подходящий под наши условия
    4. Считаем p - value и(или) доверительный интервал и делаем вывод: Если p-value < alpha - разница между группами стат.значима . Либо если доверительный интервал для разницы не включает 0.
    5. Даем (рекомендации лицам принимающим решения /принимаем решение ) выкатывать или не выкатывать новое изменение - если ключевая метрика в группе с новой фичей стат значимо лучше чем в контроле а дополнительные метрики не упали- можно раскатывать , помня про ошибку первого рода).Критерии принятия решений определяются заранее на этапе планирования эксперимента.

## ANOVA(Analysis of Variance) *Анализ дисперсии* ##
ANOVA - это способ узнать, являются ли результаты эксперимента значимыми. 

Односторонняя (**One-way**) ANOVA сравнивает два средства от буксировочной независимой группы, используя только одну независимую переменную. 

Двусторонняя (**Two-way**) ANOVA - это расширение односторонней ANOVA с использованием двух независимых переменных для расчета основного эффекта и эффекта взаимодействия.
<img src="https://github.com/allseenn/ab/blob/main/img/anova.png?raw=1" width="600">

### Chi-Square Test *Критерий хи-квадрат*###
<img src="https://github.com/allseenn/ab/blob/main/img/x2.png?raw=1" width="600">

**Chi-Square Test** Тест Хи-Квадрат проверяет, соответствует ли модель приблизительной нормальности, когда у нас есть дискретный набор точек данных.

**Goodness of Fit Test** определяет, соответствует ли выборка популяции, подходящей одной категориальной переменной к распределению. 

Тест **Хи-Квадрата для независимости** сравнивает два набора данных, чтобы увидеть, есть ли связь.



## Regression *(Регрессия)* ##
### Linear Regression *(Линейная регрессия)* ###
#### Assumptions of Linear Regression *(Допущения линейной регрессии)*####
+ Линейная зависимость
+ Многомерная нормальность
+ Нет или мало мультиколлинеарности
+ Нет или маленькая автокорреляция
+ Гомоскедастичность

**Linear Regression** *(Линейная регрессия)* — это линейный подход к моделированию взаимосвязи между зависимой переменной и одной независимой переменной. 
**An independent variable** *(Независимая переменная)* — это переменная, которая контролируется в научном эксперименте для проверки влияния на зависимую переменную. 
**A dependent variable** *(Зависимая переменная)* — это переменная, измеряемая в научном эксперименте.

<img src="https://github.com/allseenn/ab/blob/main/img/regression.webp?raw=1" width="600">

**Multiple Linear Regression** *(Множественная линейная регрессия)* — это линейный подход к моделированию отношений между зависимой переменной и двумя или более независимыми переменными.

<img src="https://github.com/allseenn/ab/blob/main/img/regression.webp?raw=1" width="600">

### Steps for Running the Linear Regression *(Этапы запуска линейной регресси)*###

<ins>Этап 1:</ins> Понимание описания модели, причинно-следственной связи и направленности

<ins>Этап 2:</ins> Проверьте данные, категориальные данные, отсутствующие данные и выбросы

+ **Outlier** *(Выброс)* — это точка данных, которая значительно отличается от других наблюдений. Мы можем использовать метод стандартного отклонения и метод межквартильного диапазона (IQR).
+ **Dummy variable** *(Фиктивная переменная)* принимает только значение 0 или 1, чтобы указать эффект для категориальных переменных.

<ins>Этап 3:</ins> **Simple Analysis** *(Простой анализ)* — Проверьте эффект сравнения между зависимой переменной с независимой переменной и независимой переменной с независимой переменной

+ Использование точечных графиков для проверки корреляции
+ **Multicollinearity** *(Мультиколлинеарность)* возникает, когда более двух независимых переменных сильно коррелированы. Мы можем использовать дисперсионный коэффициент инфляции (VIF) для измерения того, является ли VIF > 5 сильно коррелированным, и если VIF > 10, то, безусловно, существует многоколлинеарность среди переменных.

+ **Interaction Term** *(Термин взаимодействия)* подразумевает изменение наклона от одного значения к другому значению.

<ins>Этап 4:</ins> **Multiple Linear Regression** *(Множественная линейная регрессия)* — Проверьте модель и правильные переменные

<ins>Этап 5:</ins> **Residual Analysis** *(Остаточный анализ)*

+ Проверьте нормальное распределение и нормальность остатков.
+ **Homoscedasticity** *(Гомоскедастичность)* описывает ситуацию, в которой термин ошибки одинаков для всех значений независимых переменных и означает, что остатки равны по линии регрессии.

<ins>Этап 6:</ins> Интерпретация результатов регрессии

+R-Squared - это статистическая мера соответствия, которая показывает, насколько вариация зависимой переменной объясняется независимыми переменными. Более высокое значение R-квадрата представляет меньшие различия между наблюдаемыми данными и установленными значениями.
+ P-значение
+ Уравнение регрессии

## AB tests ##

### Calcs ###
https://www.evanmiller.org/ab-testing/t-test.html

https://www.rotmistrov.com/intrvl

## AB test для непрерывного распределения ##
### AB test для нормального распределения (t-test) ##
Мы провели АБ-тест на увеличение ARPU. По итогам тестирования мы получили следующие данные. Является ли результат статистически значимым с уровнем доверия 95%? Какую версию мы выкатим в продакшн ?
<table>
<tr><td></td><td>$\overline{X}$</td><td>SD</td><td>N</td></tr>
<tr><td>Variant A</td><td>300</td><td>20</td><td>600</td></tr>
<tr><td>Variant B</td><td>301</td><td>5</td><td>620</td></tr>
</table>

In [26]:
# импорт необходимых библиотек
import plotly.graph_objects as go # Для построения графиков
from scipy import stats # Для расчёта ститистик
from scipy.stats.distributions import norm # для генерации нормального распределения
mean_a = 300
mean_b = 301
sd_a = 20
sd_b = 5
size_a = 600
size_b = 620
# Генерируем выборки на основе данных
norm_a = stats.norm(loc=mean_a, scale=sd_a).rvs(size=size_a)
norm_b = stats.norm(loc=mean_b, scale=sd_b).rvs(size=size_b)
# на основен данных строим два графика
fig = go.Figure()
fig.add_trace(go.Histogram(x=norm_a, name='Norm A'))
fig.add_trace(go.Histogram(x=norm_b, name='Norm B'))
# Overlay both histograms - наслоение
fig.update_layout(barmode='overlay')
# Reduce opacity to see both histograms - прозрачность
fig.update_traces(opacity=0.5)
fig.show()

In [40]:
# Считаем статистику по выборкам
tvalue, pvalue = stats.ttest_ind(norm_a, norm_b, equal_var=False, alternative='two-sided')
md = mean_a - mean_b
se = md / tvalue
print(f"Difference between samples: {md}, standart error: {se}, p_value: {pvalue},  t_value: {tvalue}")
if pvalue < .05:
    print(f"Sample # {1 if md > 0 else 2} mean is greater")
else:
    print("No significant difference")

Difference between samples: -1, standart error: 0.4594541307950832, p_value: 0.02986601362646848,  t_value: -2.1764958305402646
Sample # 2 mean is greater


In [36]:
#Простой способ посчитать доверительный интервал
left = md - 1.96 * se
right = md + 1.96 * se
left, right

(-1.900530096358363, -0.09946990364163699)

In [42]:
# Способ без библиотек
SE = ((sd_a**2/size_a) + (sd_b**2/size_b))**0.5
l = (mean_a-mean_b) - 1.96 * SE
r = (mean_a-mean_b) + 1.96 * SE
l, r

(-2.648019991527141, 0.6480199915271412)

In [29]:
#Через scipy.stats
stats.norm.interval(.95, loc=md, scale=se)

(-1.9005135489065184, -0.09948645109348164)

In [24]:
#Через бутсрап - считаем разницу средних!
import numpy as np
m = []
booted_size = max(len(norm_a), len(norm_b))
for _ in range(1000):
    c = np.random.choice(norm_a, size=booted_size, replace=True).mean()
    t = np.random.choice(norm_b, size=booted_size, replace=True).mean()
    m.append(c-t)
np.percentile(m, [2.5, 97.5])

array([-3.1625233 ,  0.03457088])

Если интервал отрицаетльные это говорит о том, что контрольная группа А меньше тестовой группы Б, т.е. Б лучше А, принимаем Б.

Если интервал больше нуля, то контрольная группа А больше Б, т.е. А не отвергаем.

Если ноль лежит внутри интервала, то результат проведенного теста не статистически значим, т.е. мы не принимем никаой результат ни А, ни Б.

In [25]:
fig = go.Figure()
fig.add_trace(go.Histogram(x=m, name='means diff'))
fig.update_layout(barmode='overlay')
fig.update_traces(opacity=0.75)
fig.show()

### AB test для ненормальных распределений (Mann Whitney) ##
Если распределение ненорамальное, а экспонентциальное либо другое, то используем тест Манна Витни. Используем прошлые статистики.




In [12]:
# Генерируем экспоненциальное распределение на основе введенных данных
expon_a = stats.expon(loc=mean_a, scale=sd_a).rvs(size_a)
expon_b = stats.expon(loc=mean_b, scale=sd_b).rvs(size_b)
# Строим графики
fig = go.Figure()
fig.add_trace(go.Histogram(x=expon_a, name='Expon A'))
fig.add_trace(go.Histogram(x=expon_b, name='Expon B'))
# Overlay both histograms
fig.update_layout(barmode='overlay')
# Reduce opacity to see both histograms
fig.update_traces(opacity=0.75)
fig.show()

## Sample size determination (Определение размера выборки) ##
Расчет численности выборки (минимального необходимого числа юзеров) очень важный этап планирования A/B-тестов, который может с:
1. Понимать, сколько нужно юзеров для проведения эксперимента
2. Понимать сколько продлится эксперимент
3. Приоритезировать эксперименты 
### Proportions sample size (Определение выборки для доли) ###
Калькулятор - https://mindbox.ru/academy/ab-test-calculator/
используем функцию **zt_ind_solve_power()**
+ **effecitve zise** - нормированная разница между группами
+ **nobs_1** - количество наблюдейний в контроле 
+ **ratio** - отношение теста к контролю
+ **Union** - алиас для типов пайтон
+ **two-sides** - проверка двухсторонней гипотезы
Если один (только один) из этих параметров указать как None, то функция будет вычеслять именно этот параметр. Например, указав **nobes1=None**, мы посчитаем необходимое количество людей для одной выборки. Следовательно полученное значение нужно умножить на два и мы получим общее число пользователей для обеих групп.
#### Пример ####
Мы хотим провести АБ-тест баннеров. Нынешняя конверсия в клик на баннер (CTR) равна 1.5%. Мы предполагаем, что с новыми баннерами CTR вырастет до 1.7%. Определите, сколько юзеров нам нужно отправить на каждую версию с уровнем доверия 95%.
1. Для начала нам нужно вычислить effective size, который невозможно вычислить без количества наблюдений **nobs_1**, не смотря на то, что у нас задание вычислить количество наблюдений, т.е. величину выборки (замкнутый круг). Поэтому мы его заполняем сами, желательно значениями 1000 или 100, чтобы удобно было вносить конверсию. В нашем случае контрольная конверсия в переводе на целочисленные значения будет 15 (0,015), и 17 (0,017), заносим их в **conv_1** и **conv_2**. В **nobs_1** и **nobs_2** вносим по 1000, тогда в случае успеха получим конверсию 0,015 и 0,017.



In [47]:
from math import asin
def calc_propotion_effect_size(conv_1: int, nobs_1: int, conv_2: int, nobs_2: int):
    prob_1, prob_2 = conv_1 / nobs_1, conv_2 / nobs_2
    es_formula = 2  * asin(np.sqrt(prob_1)) - 2 * asin(np.sqrt(prob_2)) # Математическая формула
    return es_formula

conv_1, conv_2 = 15, 17
nobs_1, nobs_2 = 1000, 1000
prob_1, prob_2 = conv_1 / nobs_1, conv_2 / nobs_2 #0.015, 0.017
print(prob_1, prob_2)
calc_propotion_effect_size(conv_1=conv_1, nobs_1=nobs_1, conv_2=conv_2, nobs_2=nobs_2)

0.015 0.017


-0.015947131645016377

2. Полученный результат effective size = -0.015947131645016377 передается в функцию для дальнейших расчетов. Не забывает импортировать zt_ind_solve_power из библиотеки statsmodels.stats.power и Union для создания нового типа данных, способного принимать float, int.

In [52]:
from statsmodels.stats.power import zt_ind_solve_power
from typing import Union
def calc_sample_size_proportion(effect_size: float,
                                alpha: float = .05,
                                beta: float = .2,
                                ratio: Union[float, int] = 1):
    
    n = zt_ind_solve_power(effect_size=effect_size,
                           alpha=alpha,
                           power=(1 - beta),
                           ratio=ratio,
                  )
    return int(n * 2)
effect_size = calc_propotion_effect_size(conv_1=conv_1, nobs_1=nobs_1, conv_2=conv_2, nobs_2=nobs_2)
calc_sample_size_proportion(effect_size=effect_size, alpha=.05, beta=.2)

123452

Ответ: 123452 пользователя необходимо для привлечения к AB тестированию при альфе 5% и мощности теста 80% в обеих группах.


Теперь поиграим с нашими параметрами, подставляя None, например вместо альфы


In [56]:
effect_size = calc_propotion_effect_size(conv_1=conv_1, nobs_1=nobs_1, conv_2=conv_2, nobs_2=nobs_2)
alpha = .05
beta = .2 # power = None
nobs1 = 61726
ratio = None
zt_ind_solve_power(effect_size=effect_size,
                           alpha=alpha,
                           power=(1 - beta), # power = power
                           nobs1 = nobs1,
                           ratio=ratio)

1.000015973591064

### Mean sample size (Определение численности выборки для среднего) - непрерывная величина ###
https://statulator.com/SampleSize/ss2M.html
#### Пример ####
Дизайнер предложил добавить рекомендательную систему на этапе формирования корзины. Нынешний средний чек равен 2167 рублей, а стандартная ошибка (SD) равна 69. Мы предполагаем, что в новой версии средний чек вырастет 2180. Определите, сколько юзеров нам нужно отправить на каждую версию с уровнем доверия 95%.


In [58]:
def calc_continuous_effect_size(mean_1: Union[float, int],
                                std_1: Union[float, int],
                                mean_2: Union[float, int],
                                std_2: Union[float, int],
                                nobs_1: int = 10_000,
                                nobs_2: int = 10_000):
    
    es_formula = (mean_1 - mean_2) / ((std_1**2 + std_2**2) / 2) ** 0.5 # Математическая формула    
    return es_formula

mu_control, mu_test = 2167, 2180
std_control, std_test = 69, 69 * 1.15 # Второе СКО поменяем в большую сторону, чтобы они билы разными и приближенными к реальной жизни

calc_continuous_effect_size(mean_1=mu_control, std_1=std_control, mean_2=mu_test, std_2=std_test) #-0.18840579710144928

-0.17483621585265752

In [61]:
from statsmodels.stats.power import tt_ind_solve_power
def calc_sample_size_continuous(effect_size: float,
                                alpha: float = .05,
                                beta: float = .2,
                                ratio: Union[float, int] = 1):
    
    n = tt_ind_solve_power(effect_size=effect_size,
                           alpha=alpha,
                           power=(1 - beta),
                           ratio=ratio,
                  )
    return int(n * 2)
effect_size = calc_continuous_effect_size(mean_1=mu_control, std_1=std_control, mean_2=mu_test, std_2=std_test)
calc_sample_size_continuous(effect_size=effect_size, alpha=.05, beta=.2)

1029

Ответ: необходимо 1029 юзера в двух группах в сумме, чтоб провести АБ тестирование непрерывной величины в виде среднего чека.